pkg/connector/ipa_object_cache.go

```diff
@@ -0,0 +1,80 @@
+package connector
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/conductorone/baton-ipa/pkg/ldap"
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	ldap3 "github.com/go-ldap/ldap/v3"
+)
+
+type ipaObjectCache struct {
+	client *ldap.Client
+	baseDN *ldap3.DN
+
+	entries map[string]*ipaObject
+	mu      sync.RWMutex
+}
+
+type ipaObject struct {
+	ipaUniqueID  string
+	dn           string
+	resourceType *v2.ResourceType
+}
+
+func (c *ipaObjectCache) get(ctx context.Context, dn string) (*ipaObject, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	m, ok := c.entries[dn]
+	if ok {
+		return m, nil
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	m, ok = c.entries[dn]
+	if ok {
+		return m, nil
+	}
```

Comment:
I thing this might cause a deadlock, since the read will be released in when the function finish

<img width="864" alt="image" src="https://github.com/user-attachments/assets/ac70ac7b-2443-4f20-b71e-9d9254b3f62c" />